### DISCUSSION TITLE
Unable to access self-implemented "abstract" interface members.

### DISCUSSION THREAD

**Discussion by nathan130200** [dotnet/csharplang#7992] [Q&A] [answered]
Why i cannot access self-implemented members in "abstract" interfaces.

Example:

```cs
public interface IBase64Content
{
    public string ContentAsBase64
    {
        get => (this as XElement).Value.FromBase64();
        set => (this as XElement).Value = value.ToBase64();
    }
}

```

When implementing that interface:

```cs
public class Response : XElement, IXmppFactoryInfo, IBase64Content
{
    static IEnumerable<XName> IXmppFactoryInfo.GetNames()
    {
        yield return SaslNames.Response;
    }

    public Response() : base(SaslNames.Response)
    {

    }
}
```

This will not compile:

```
var resp = new Response();
resp.ContentAsBase64 = "should encode as base64 here";
```

`CS1061: 'Response' does not contain a definition for ContentAsBase64.`

Instead i MUST do a explicit cast to interface type to use property `ContentAsBase64` don't make sense?

**Comment by FaustVX** [comment, marked as answer]
It is intended to be used like that.
DIM **does not** exists on the implementing `class`, only on the `interface`.
https://learn.microsoft.com/en-us/dotnet/csharp/language-reference/proposals/csharp-8.0/default-interface-methods#concrete-methods-in-interfaces

**Reply by nathan130200** [reply]
This is a little sad, because I'll have to copy and paste the same code several times, because in the context i am working with, the class can only extend to XElement and C# doesn't allow multiple subclasses (something like C++ does).

**Reply by FaustVX** [reply]
You could write 2 extension methods (`GetContentAsBase64` and `SetContentAsBase64`) on `XElement`

**Reply by nathan130200** [reply]
Yes for sure, I just thought that the self-implement members in the interface worked like that, so I tried to see if I could stop using extension methods in this case. In this context, it is really worth using extension methods.

**Comment by HaloFour** [comment]
Interface members aren't inherited like class members are.  The members only exist through the interface, unless the type also declares the members, which is the typical case.  But with default or explicit implementation, the interface member effectively does not exist on the type and cannot be called directly without calling through the interface member.

There [was a proposal](https://github.com/dotnet/csharplang/issues/2337) to add a `base(T).M` to call the interface member directly, but [that proposal had kinda stalled](https://github.com/dotnet/csharplang/discussions/7227#discussioncomment-6002659).

**Reply by nathan130200** [reply]
Ah yes, I thought the compiler could understand and resolve automatically when I invoked the interface member
